base/src/dma/dma.cpp

```diff
@@ -0,0 +1,259 @@
+#include <cstdint>
+
+#include "dma.hpp"
+#include "../system/stm32h750xx.h"
+#include "../system/core_cm7.h"
+
+using namespace dma;
+
+#define DMA_CHANNEL_OFFSET(mChannelNum) 6*(mChannelNum-1) // word indexing
+
+///
+/// Constructor
+///
+DmaController::DmaController(uint8_t controller_num):
+	mControllerHwNum(controller_num)
+	{
+	for (uint8_t iChannel = 0; iChannel < DMA1_NUM_CHANNELS; iChannel++){
+		mChannels[iChannel].mChannelNum = iChannel + 1;
+		mChannels[iChannel].mChannelClaimed = 0;
+		mChannels[iChannel].mChannelController = this;
+		switch(mControllerHwNum){
+			case 1:
+				mChannels[iChannel].mStreamHw = DMA1_Stream0+iChannel;
+				mControllerHw = DMA1;
+				break;
+			case 2:
+				mChannels[iChannel].mStreamHw = DMA2_Stream0+iChannel;
+				mControllerHw = DMA2;
+				break;
+		}
+	}
+
+	RCC->AHB1ENR |= 1 << (mControllerHwNum -1); // enable DMA controller
+	RCC->AHB1RSTR |= 1 << (mControllerHwNum -1); // reset DMA controller
+	RCC->AHB1RSTR &= ~(1 << (mControllerHwNum -1)); // reset DMA controller
+}
+
+///
+/// Set the peripheral request for a specific dma channel.
+/// @param perID Request ID
+/// @param channel channel
+///
+void DmaController::setRequest(uint8_t perID, uint8_t channel){
+	perID &= 0x3F; // keep it to 6 bits.
+
+	*(&(DMAMUX1_Channel0->CCR)+(channel-1) + ((mControllerHwNum-1)*8)) = perID;
+}
+
+
+///
+/// Claim a DMA channel.
+/// @param channelNum Channel number to claim (1-8)
+/// @return Pointer to Dma channel object or NUll if channel not available.
+///
+DmaController::DmaChannel* DmaController::claimChannel(uint8_t channelNum){
+	if(channelNum < 1 || channelNum > DMA1_NUM_CHANNELS){
+		return NULL;
+	}
+	if(mChannels[channelNum-1].mChannelClaimed == 0){
+		mChannels[channelNum-1].claim();
+		return &mChannels[channelNum-1];
+	}
+	return NULL;
+}
+
+///
+/// Release a DMA channel.
+/// @param channel Pointer to channel to free
+///
+void DmaController::releaseChannel(DmaController::DmaChannel* channel){
+	if(!channel){
+		return; // Null pointer
+	}
+	channel->release();
+}
+
+///
+/// Get pointer to the next available DMA channel or NULL if all are taken
+/// @return Next availble channel or NULL if none available
+///
+DmaController::DmaChannel* DmaController::getAvailableChannel(){
```

Comment:
This function looks like it claims a DMA channel. Might be good to a comment saying that or have it return just the channel number

Reply:
Modified function name to claimAvailableChannel and updated documentation to specifically say that it claims the channel.